beacon-chain/forkchoice/doubly-linked-tree/store.go

```diff
@@ -62,20 +62,32 @@ func (s *Store) head(ctx context.Context) ([32]byte, error) {
 
 // insert registers a new block node to the fork choice store's node list.
 // It then updates the new node's parent with the best child and descendant node.
-func (s *Store) insert(ctx context.Context,
+func (s *Store) insert(
+	ctx context.Context,
 	slot primitives.Slot,
-	root, parentRoot, payloadHash [fieldparams.RootLength]byte,
-	justifiedEpoch, finalizedEpoch primitives.Epoch) (*Node, error) {
+	root, parentRoot, payloadHash, parentHash [fieldparams.RootLength]byte,
+	justifiedEpoch, finalizedEpoch primitives.Epoch,
+) (*Node, error) {
 	ctx, span := trace.StartSpan(ctx, "doublyLinkedForkchoice.insert")
 	defer span.End()
 
 	// Return if the block has been inserted into Store before.
-	if n, ok := s.nodeByRoot[root]; ok {
-		return n, nil
+	n, rootPresent := s.nodeByRoot[root]
+	m, hashPresent := s.nodeByPayload[payloadHash]
+	if rootPresent {
+		if payloadHash == [32]byte{} {
+			return n, nil
+		}
+		if hashPresent {
+			return m, nil
+		}
```

Comment:
Here we assume no root collisions. In pre-epbs conditions we would have both `rootPresent` and `hashPresent` on existing nodes pointing to the same node, so here `n,nil` or `m,nil` is the same return. 

On `ePBS` there are different outcomes here: 
- When importing a new block,  if we try to do it twice, we would have already the `rootPresent` but not `hashPresent`. We check here that the incoming `payloadHash` is zero so that we return the already imported empty node. 
- If on the other hand we are trying to import a new payload, if the root already existed but the hash did not, that is not a problem, we create a new node.
- There is some bad behavior here if we had `hashPresent == true` but `rootPresent == false`. This can only happen with either hash collisions or `payloadHash == [32]byte{}`.